File: MicroTDCore/Sources/MicroTDCore/Definitions/RelicDef.swift
// RelicDef.swift
// Relic definitions for roguelite choices

import Foundation

/// Effect modifier for relics
public struct RelicEffect: Codable, Equatable {
    public enum EffectType: String, Codable {
        case towerDamageMultiplier
        case towerRangeMultiplier
        case towerFireRateMultiplier
        case enemySlowOnHit
        case coinMultiplier
        case startingCoins
    }
    
    public let type: EffectType
    public let value: Double
    
    public init(type: EffectType, value: Double) {
        self.type = type
        self.value = value
    }
}

/// Relic rarity tiers
public enum RelicRarity: String, Codable {
    case common
    case uncommon
    case rare
    case legendary
}

/// Relic definition
public struct RelicDef: Codable, Equatable {
    public let id: String
    public let name: String
    public let description: String
    public let rarity: RelicRarity
    public let effects: [RelicEffect]
    
    public init(id: String, name: String, description: String, rarity: RelicRarity, effects: [RelicEffect]) {
        self.id = id
        self.name = name
        self.description = description
        self.rarity = rarity
        self.effects = effects
    }
}

/// Container for all relic definitions
public struct RelicDefinitions: Codable {
    public let relics: [RelicDef]
    
    public init(relics: [RelicDef]) {
        self.relics = relics
    }
    
    public func relic(withID id: String) -> RelicDef? {
        return relics.first { $0.id == id }
    }
    
    public func relics(withRarity rarity: RelicRarity) -> [RelicDef] {
        return relics.filter { $0.rarity == rarity }
    }
}

File: expo/src/core/definitions/RelicDef.ts
// RelicDef.ts
// Relic type definitions — ported from MicroTDCore/Definitions/RelicDef.swift

/**
 * Relic effect types
 */
export type EffectType =
    | 'towerDamageMultiplier'
    | 'towerRangeMultiplier'
    | 'towerFireRateMultiplier'
    | 'coinMultiplier'
    | 'startingCoins'
    | 'enemySlowOnHit';

/**
 * Single effect applied by a relic
 */
export interface RelicEffect {
    type: EffectType;
    value: number;
}

/**
 * Relic rarity tiers
 */
export type RelicRarity = 'common' | 'uncommon' | 'rare' | 'legendary';

/**
 * Relic definition (from JSON or hardcoded)
 */
export interface RelicDef {
    id: string;
    name: string;
    description: string;
    rarity: RelicRarity;
    effects: RelicEffect[];
}

/**
 * Container for all relic definitions
 */
export class RelicDefinitions {
    constructor(public readonly relics: RelicDef[]) { }

    /**
     * Find relic by ID
     */
    relic(id: string): RelicDef | undefined {
        return this.relics.find(r => r.id === id);
    }

    /**
     * Get all relics of a specific rarity
     */
    relicsByRarity(rarity: RelicRarity): RelicDef[] {
        return this.relics.filter(r => r.rarity === rarity);
    }
}
